pkg/kubelet/kuberuntime/kuberuntime_container_linux.go

```diff
@@ -107,17 +107,22 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
 
 	lcr.HugepageLimits = GetHugepageLimitsFromResources(container.Resources)
 
-	if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
-		// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
-		// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
-		switch m.memorySwapBehavior {
-		case kubelettypes.LimitedSwap:
-			swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
-		default:
-			swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
+	// swap is only configured if a swap controller is available.
+	if swapControllerAvailable() {
+		if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+			// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
+			// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
+			switch m.memorySwapBehavior {
+			case kubelettypes.LimitedSwap:
+				swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
+			default:
+				swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
+			}
+		} else {
+			swapConfigurationHelper.ConfigureNoSwap(lcr)
 		}
-	} else {
-		swapConfigurationHelper.ConfigureNoSwap(lcr)
+	} else if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+		klog.ErrorS(errors.New("no cgroup swap controller present"), "ignoring NodeSwap feature", "pod", klog.KObj(pod), "containerName", container.Name)
 	}
```

Comment:
Is there some way to cleanup these if/elses to read more nicely. In partifular, this is harad to look at:
```
if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
```

What about:
```
	if !swapControllerAvailable() && !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		// Nothing to do
	}
	if !swapControllerAvailable() && utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		klog.ErrorS(errors.New("no cgroup swap controller present"), "ignoring NodeSwap feature", "pod", klog.KObj(pod), "containerName", container.Name)
	}
	if swapControllerAvailable() && !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo);
		swapConfigurationHelper.ConfigureNoSwap(lcr)
	}
	if swapControllerAvailable() && utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo);
		// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
		// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
		switch m.memorySwapBehavior {
		case kubelettypes.LimitedSwap:
			swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
		default:
			swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
		}
	}
```

Reply:
I was wondering about the nesting too. What about adding an `addSwapResources` method that we call here instead.

```
func (m *kubeGenericRuntimeManager) addSwapResources(...) {
	if !swapControllerAvailable() && !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		return
	}
	if !swapControllerAvailable() {
		klog.ErrorS(errors.New("no cgroup swap controller present"), "ignoring NodeSwap feature", "pod", klog.KObj(pod), "containerName", container.Name)
		return
	}
	swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo)
	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		swapConfigurationHelper.ConfigureNoSwap(lcr)
		return
	}
	// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
	// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
	swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo)
	switch m.memorySwapBehavior {
	case kubelettypes.LimitedSwap:
		swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
	default:
		swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
	}
}
```

Reply:
definitely seems cleaner to me

Reply:
@klueska I reworked this a bit with quick-returns where applicable. If you really want me to check the feature gate on all if statements, I can do so.